### PR TITLE
smtp: Fix bcc being ignored

### LINKED ISSUE #127: smtp: bcc addresses are not included when sending email
As of [RFC2822](https://tools.ietf.org/html/rfc2822#section-3.6.3) and [this SO answer](https://stackoverflow.com/a/26611044/6093458), BCC lists may be stripped out on "DATA" command and should be looked up on the "RCPT" messages instead of looking for them only on the "DATA" payload.

### PR DESCRIPTION
**Pull request by mrvik**
Some clients don't send BCC on message body (this behavior is allowed by [RFC 2822](https://tools.ietf.org/html/rfc2822#section-3.6.3) section 3.6.3) and thus, the BCC list may not appear there.

Commits here compare RCPT messages and addresses received on DATA so addresses not present on DATA are handled as blind copies.
If mail client sends BCC on DATA, this additions are ignored (handler uses provided BCC instead of falling-back to generating one)

Fixes #127

### CODE REVIEW

File: smtp/smtp.go [hunk 1]
```diff
@@ -38,18 +38,50 @@ func formatHeader(h mail.Header) string {
...
 
-func (s *session) Rcpt(to string) error {
-	return nil
+func (s *session) Rcpt(to string) (err error) {
+	log.Printf("Sending mail for %s", to)
+	if to == "" {
+		return
+	}
+
+	// Seems like github.com/emersion/go-smtp/conn.go:487 removes marks on message
```
Comment (emersion) [inline smtp/smtp.go:58]: `mail.ParseAddress` should be able to parse the string here, no need to do it manually.

File: smtp/smtp.go [hunk 2]
```diff
@@ -38,18 +38,50 @@ func formatHeader(h mail.Header) string {
...
+	name, to := strings.Join(split[:ln-1], " "), split[ln-1]
+
+	s.allReceivers = append(s.allReceivers, &mail.Address{
+		Name:    name,
+		Address: to,
+	})
+	return
+}
+
+func (s *session) bccFromRest(ignoreMails []*mail.Address) (final []*mail.Address) {
```
Comment (emersion) [inline smtp/smtp.go:70]: No need to use naked returns here. See https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: smtp/smtp.go [hunk 3]
```diff
@@ -38,18 +38,50 @@ func formatHeader(h mail.Header) string {
...
+	ignore := make(map[string]struct{})
+	for _, mail := range ignoreMails {
+		ignore[mail.Address] = struct{}{}
+	}
+
+	final = make([]*mail.Address, 0, len(s.allReceivers))
+	var exists bool
```
Comment (emersion) [inline smtp/smtp.go:77]: No need to define this variable here, can use `:=` in the assignment instead.

Comment (mrvik) [reply smtp/smtp.go:77]: Bruh, thought declaring vars on loops was a bad practice, but just googled it and lecturer was wrong. Changing it now, thanks

File: smtp/smtp.go (hunk 1 above)
Comment (mrvik) [reply smtp/smtp.go:58]: Yes, it works, but needs changes on `go-smtp`.
Consider this playground: https://play.golang.org/p/LegwsnQbpos

Comment (emersion) [reply smtp/smtp.go:58]: Hm, go-smtp isn't supposed to return the `TO:` part…

Comment (emersion) [reply smtp/smtp.go:58]: This is supposed to work: https://play.golang.org/p/qMBIcNqhvox

If it doesn't, this is a go-smtp bug.

**Comment by mrvik** [comment]
Added another commit because the `bccList` check happens after another check to ensure there are recipients, and mails can be sent only with BCC addresses

File: smtp/smtp.go [hunk 4]
```diff
@@ -38,20 +38,46 @@ func formatHeader(h mail.Header) string {
...
 	return nil
 }
 
 func (s *session) Rcpt(to string) error {
+	if to == "" {
+		return nil
+	}
+
+	// Seems like github.com/emersion/go-smtp/conn.go:487 removes marks on message
+	// "to" is added into allReceivers blindly
```
Comment (emersion) [inline smtp/smtp.go:58]: I don't think either of these two comments are really important.

**Comment by emersion** [comment]
It looks like this doesn't build: https://builds.sr.ht/~emersion/job/350538#task-build-83

**Comment by emersion** [comment]
The logic LGTM. When you've fixed the build, can you squash everything into a single commit? Thanks.

**Review by mrvik — COMMENTED** [review]
> The logic LGTM. When you've fixed the build, can you squash everything into a single commit? Thanks.

Ok, will do

**Comment by mrvik** [comment]
Seems like b116e0d5b0a8868599584266e53c77a5e93fbeb7 was the commit that broke the build. Instead of `"charset": "utf-8"` there is `"charset", "utf-8"`

**Comment by emersion** [comment]
Ooops,. sorry about that! Fixing it now.

**Comment by emersion** [comment]
358ca10548fbac7f695d590dfa6882675b48b757

**Comment by mrvik** [comment]
Haha, I'll rebase it and push as a single commit

**Review by emersion — APPROVED** [review]
Thanks!

**Comment by mrvik** [comment]
Thanks to you, I love this project, keep it going

**Pull request merged**